gh-116560: Add PyLong_GetSign() public function (#116561)
python/cpython: Doc/c-api/long.rst

   .. versionadded:: 3.13

.. c:function:: int PyLong_Sign(PyObject *obj)

   Return ``-1``, ``0`` or ``+1`` if the integer object *obj* is negative, zero
   or positive, respectively.

>    Returns ``-1`` on error.  Use :c:func:`PyErr_Occurred` to disambiguate.

Reviewer: This API is bad. Please change the public API to: `int PyLong_Sign(PyObject *obj, int *result)`.

* Set `*result` and return 0 on success.
* Return -1 on error.

This way, we don't have to check for PyErr_Occurred().

See C API guidelines: https://devguide.python.org/developer-workflow/c-api/index.html#guidelines-for-expanding-changing-the-public-api

Author: Hmm, I'm not sure that this is against the C API guidline.  ("Most other API must return -1 with an exception raised on error, and 0 on success.", etc)

With the new design, 1) ``_PyLong_Sign()`` should be restored, 2) we loose ability to do something like:
```c
if (PyLong_Sign(its_surely_a_long_int) < 0) {
    <do stuff>;
}
```
Maybe we should instead document all cases, when an exception could be raised?  (It's easy: ``NULL`` pointer or not an ``int`` subclass.)

In this way, users could choose if they actually want to call ``PyErr_Occurred()`` or be sure this can't happen in their code.  Most consumers of the API (e.g. gmpy2), probably, will fall to the second camp (and just remove ``_`` prefix).

Reviewer: > Most other API must return -1 with an exception raised on error, and 0 on success.

Yes, new API functions have to follow this guideline. The result is an argument passed by pointer:

```c
int sign;
if (PyLong_Sign(obj, &sign) < 0) { ... handle error ... }
... use sign ...
```

> Maybe we should instead document all cases, when an exception could be raised? (It's easy: NULL pointer or not an int subclass.)

This code fails obviously (go to "handle error") if obj is not an integer. For example, if obj is a string.

Usually, we don't document error cases, since it would hard to maintain this in the documentation. It's better to document the case which don't fail, like "obj must be an Python int object".

> In this way, users could choose if they actually want to call PyErr_Occurred() or be sure this can't happen in their code.

The chosen guideline is that all C API consumers must check for errors. If you know for sure that the function **cannot** fail, just ignore the function result and use `sign`.

Reviewer: The problem of the current API is that check for `PyErr_Occurred()` is inefficient, whereas if we design the API differently, `PyErr_Occurred()` doesn't have to be called: it's faster.

Author: I hope, I did this.

   .. versionadded:: 3.13

.. c:function:: int PyLong_GetSign(PyObject *obj, int *sign)

>    Retrieve the sign of integer object *obj* (``0``, ``-1`` or ``+1`` for zero,
>    negative or positive integer, respectively) in a variable *sign*.

Reviewer: ```suggestion
   Get the sign of the integer object *obj*: ``0``, ``-1`` or ``+1`` for zero,
   negative or positive integer, respectively.
```

>    Return ``0`` on success, else ``-1`` with an exception set.  This function
>    always succeeds if *obj* is a :c:type:`PyLongObject` or its subtype.

Reviewer: Your doc doesn't mention sign. I suggest (more or less a copy/paste of C API WG doc):

```suggestion
   * On success, set *\*sign* to the integer sign, and return 0.
   * On failure, set an exception, and return -1.

   This function always succeeds if *obj* is a :c:type:`PyLongObject`
   or its subtype.
```

>    Get the sign of the integer object *obj*: ``0``, ``-1`` or ``+1`` for zero,
>    negative or positive integer, respectively.
> 
>    Return ``0`` on success, else ``-1`` with an exception set.  This function
>    always succeeds if *obj* is a :c:type:`PyLongObject` or its subtype.

Author: ```suggestion
   Get the sign of the integer object *obj*.
   
   On success, set *\*sign* to the integer sign  (0, -1 or +1 for zero, negative or
   positive integer, respectively) and return 0.

   On failure, return -1 with an exception set.  This function always succeeds
   if *obj* is a :c:type:`PyLongObject` or its subtype.
```

>    .. versionadded:: 3.13

Reviewer: ```suggestion
   .. versionadded:: 3.14
```
